### PR TITLE
feat: Allow secret references in KeycloakRealmComponent

### PR DESCRIPTION
**Pull request by regiontog**
## Description
Allow the config property of KeycloakRealmComponent to have references to secret values.

In order to protect the AD service account used to integrate keycloak with AD as a User Federation a way to set realm component config values from k8s secrets is needed. There are a few ways to enable this, but I saw something similar was done for KeycloakRealmIdentityProvider so I tried to keep this as similar as possible.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?
I made some unit and integration tests to trigger failures when referring to secret values and ran `TEST_KEYCLOAK_URL="http://localhost:8080" make test` while developing the changes.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Pull Request contain one commit. I squash my commits

## Additional context
First time working with golang so I'm sure there are some improvements to make on this PR. I wasn't sure if this is the norm, but I ran `mockery` to generate a new `ref_mock.go` file. I couldn't find any relevant documentation to change, and I'm not quite sure where to look for "new warnings" but I didn't see anything while working on this.

### CODE REVIEW

File: api/v1/keycloakcomponent_types.go [hunk 1]
```diff
@@ -33,6 +33,9 @@ type KeycloakComponentSpec struct {
 	ParentRef *ParentComponent `json:"parentRef,omitempty"`
 
 	// Config is a map of component configuration.
+	// Map key is a name of configuration property, map value is an array value of configuration properties.
```
Comment (zmotso) [inline api/v1/keycloakcomponent_types.go:36]: Run `make manifests` to update CRD

File: controllers/keycloakrealmcomponent/keycloakrealmcomponent_controller_integration_test.go [hunk 2]
```diff
@@ -273,4 +274,79 @@ var _ = Describe("KeycloakRealmComponent controller", func() {
...
+			createdComponent := &keycloakApi.KeycloakRealmComponent{}
+			err := k8sClient.Get(ctx, types.NamespacedName{Name: component.Name, Namespace: ns}, createdComponent)
+			if err != nil {
+				return false
+			}
+
+			return createdComponent.Status.Value == helper.StatusOK
+		}, timeout, interval).Should(BeTrue())
+	})
+	It("Should fail to create resource with nonexistant secret reference in config", func() {
```
Comment (zmotso) [inline controllers/keycloakrealmcomponent/keycloakrealmcomponent_controller_integration_test.go:320]: please fix typo `nonexistant`

File: controllers/keycloakrealmcomponent/keycloakrealmcomponent_controller.go [hunk 3]
```diff
@@ -184,6 +190,10 @@ func (r *Reconcile) tryReconcile(
 		return fmt.Errorf("unable to create keycloak component: %w", err)
 	}
 
+	if err = r.secretRefClient.MapComponentConfigSecretsRefs(ctx, keycloakComponent.Config, keycloakRealmComponent.Namespace); err != nil {
```
Comment (zmotso) [inline controllers/keycloakrealmcomponent/keycloakrealmcomponent_controller.go:193]: Let's modify `createKeycloakComponent`func to not modify the original object keycloakRealmComponent.

```go
func (r *Reconcile) createKeycloakComponent(
	ctx context.Context,
	component *keycloakApi.KeycloakRealmComponent,
	kcRealmName string,
	kClient keycloak.Client,
) (*adapter.Component, error) {
	ksComponent := &adapter.Component{
		Name:         component.Spec.Name,
		Config:       make(map[string][]string),
		ProviderID:   component.Spec.ProviderID,
		ProviderType: component.Spec.ProviderType,
	}

	for k, v := range component.Spec.Config {
		ksComponent.Config[k] = make([]string, len(v))
		copy(ksComponent.Config[k], v)
	}
	...
```

File: api/v1/keycloakcomponent_types.go (hunk 1 above)
Comment (zmotso) [inline api/v1/keycloakcomponent_types.go:36]: Change the commit message to our approach:
`feat: Allow secret references in KeycloakRealmComponent (#issue_number)`
Example: feat: Allow secret references in KeycloakRealmComponent (#123)

**Review by zmotso — COMMENTED** [review]
Thanks, @regiontog, for your contribution.
Please fix a couple of comments, and we will merge this PR.

File: api/v1/keycloakcomponent_types.go (hunk 1 above)
Comment (regiontog) [reply api/v1/keycloakcomponent_types.go:36]: Done

File: controllers/keycloakrealmcomponent/keycloakrealmcomponent_controller.go (hunk 3 above)
Comment (regiontog) [reply controllers/keycloakrealmcomponent/keycloakrealmcomponent_controller.go:193]: Done

File: api/v1/keycloakcomponent_types.go (hunk 1 above)
Comment (regiontog) [reply api/v1/keycloakcomponent_types.go:36]: I created issue #30 to have something to refer to. Now reflected in commit message

**Comment by regiontog** [comment]
@zmotso Thank you for taking a look at my PR and the prompt feedback! I think I have addressed all of your comments now.

**Comment by regiontog** [comment]
Fixed validate-docs(I hope) and unit tests. I was not entirely sure if you would prefer to keep the old behavior of allowing Config to be a nil value or changing the test to construct an empty map. But I went for the latter since you suggested that code.

**Comment by SergK** [comment]
@regiontog Thank you

**Pull request merged**